### PR TITLE
In Sliding Sync, find out which rooms we have left from state events

### PR DESCRIPTION
**Pull request by andybalaam**
Fixes https://github.com/matrix-org/matrix-rust-sdk/issues/1834
(I hope. Previously, we were not recognising that we had left a room when using sliding sync, and with this change, we should be.)

I suggest reviewing commits separately. One is the functional change and one is a performance improvement.

Look through m.room.member events when processing Sliding Sync responses, finding those that refer to our user's membership, so that we can recognise which rooms we have left. (I confirmed in the Sliding Sync matrix room that this is the way we should recognise which rooms we have left.)

### CODE REVIEW

File: crates/matrix-sdk-base/src/rooms/normal.rs [hunk 1]
```diff
@@ -83,6 +83,19 @@ pub enum RoomState {
     Invited,
 }
 
+impl From<&MembershipState> for RoomState {
+    fn from(membership_state: &MembershipState) -> Self {
+        match membership_state {
+            MembershipState::Ban => Self::Left, // TODO: is this right?
```
Comment (Hywan) [inline crates/matrix-sdk-base/src/rooms/normal.rs:89]: I guess it's fine. cc @manuroe?

File: crates/matrix-sdk-base/src/rooms/normal.rs [hunk 2]
```diff
@@ -83,6 +83,19 @@ pub enum RoomState {
...
+            MembershipState::Invite => Self::Invited,
+            MembershipState::Join => Self::Joined,
+            MembershipState::Knock => Self::Left, // TODO: is this right?
```
Comment (Hywan) [inline crates/matrix-sdk-base/src/rooms/normal.rs:92]: I guess it's fine. cc @manuroe?

File: crates/matrix-sdk-base/src/sliding_sync.rs [hunk 3]
```diff
@@ -305,15 +308,55 @@ impl BaseClient {
...
+    pub(crate) async fn handle_own_room_membership(
+        &self,
+        required_state: &[Raw<AnySyncStateEvent>],
+        room_info: &mut RoomInfo,
+    ) {
+        // Note: we deserialise these same events inside handle_state(). We could refactor to do
+        // this only once.
+
+        for raw_event in required_state {
+            let event = match raw_event.deserialize() {
```
Comment (Hywan) [inline crates/matrix-sdk-base/src/sliding_sync.rs]: cc @jplatte: Is there an optimisation we can do here?

File: crates/matrix-sdk-base/src/sliding_sync.rs [hunk 4]
```diff
@@ -425,6 +471,41 @@ mod test {
...
+        client.process_sliding_sync(&response).await.expect("Failed to process sync");
+        // (sanity: state is join)
+        assert_eq!(client.get_room(room_id).unwrap().state(), RoomState::Joined);
+
+        // And then leave...
+        let mut room = v4::SlidingSyncRoom::new();
+        set_room_left(&mut room, user_id);
+        let response = response_with_room(room_id, room).await;
+        client.process_sliding_sync(&response).await.expect("Failed to process sync");
+        // (sanity: state is invite)
```
Comment (Hywan) [inline crates/matrix-sdk-base/src/sliding_sync.rs:500]: Hmm, `state is _left_` no?

**Review by Hywan — CHANGES_REQUESTED** [review]
Thanks for the patch and for the tests, they are great!
One small modification, but apart from that, it's fine.

File: crates/matrix-sdk-base/src/rooms/normal.rs (hunk 1 above)
Comment (jplatte) [reply crates/matrix-sdk-base/src/rooms/normal.rs:89]: Looks right to me.

File: crates/matrix-sdk-base/src/rooms/normal.rs (hunk 2 above)
Comment (jplatte) [reply crates/matrix-sdk-base/src/rooms/normal.rs:92]: Same here.

File: crates/matrix-sdk-base/src/sliding_sync.rs (hunk 3 above)
Comment (jplatte) [reply crates/matrix-sdk-base/src/sliding_sync.rs]: Yes, we could do an `if raw_event.get_field::<StateEventType>("type").is_ok_and(|t| t == Some(StateEventType::RoomMember))` and if we enter that branch, deserialize using `raw_event.deserialize_as::<SyncRoomMemberEvent>()`. That way we never deserialize non-member events in this function.

Comment (andybalaam) [reply crates/matrix-sdk-base/src/sliding_sync.rs]: Sorry, this was confusing: my second commit avoids double-deserialising by doing it outside of this function.

If you're happy with that, we don't need to do anything else I think.

File: crates/matrix-sdk-base/src/sliding_sync.rs (hunk 4 above)
Comment (andybalaam) [reply crates/matrix-sdk-base/src/sliding_sync.rs]: Thanks! Fixed in 22fd44df10

**Review by Hywan — APPROVED** [review]

**Pull request merged**